File: php/tests/IntegrationTest.php
<?php
/*
 * ScratchBird-driver
 * Copyright (c) 2025-2026 Dalton Calford
 *
 * Licensed under the Initial Developer's Public License Version 1.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at:
 * https://www.firebirdsql.org/en/initial-developer-s-public-license-version-1-0/
 */

use PHPUnit\Framework\TestCase;
use ScratchBird\PDO\ScratchBirdPDO;

final class IntegrationTest extends TestCase
{
    public function testSelect(): void
    {
        $dsn = getenv('SCRATCHBIRD_PHP_URL');
        if (!$dsn) {
            $this->markTestSkipped('SCRATCHBIRD_PHP_URL not set');
        }
        $pdo = new ScratchBirdPDO($dsn);
        $stmt = $pdo->query('SELECT 1');
        $row = $stmt->fetch(\PDO::FETCH_NUM);
        $this->assertSame(1, (int)$row[0]);
    }

    public function testPrepareBind(): void
    {
        $dsn = getenv('SCRATCHBIRD_PHP_URL');
        if (!$dsn) {
            $this->markTestSkipped('SCRATCHBIRD_PHP_URL not set');
        }
        $pdo = new ScratchBirdPDO($dsn);
        $stmt = $pdo->prepare('SELECT ?::INTEGER');
        $stmt->execute([42]);
        $row = $stmt->fetch(\PDO::FETCH_NUM);
        $this->assertSame(42, (int)$row[0]);
    }

    public function testTypesFixture(): void
    {
        $dsn = getenv('SCRATCHBIRD_PHP_URL');
        if (!$dsn) {
            $this->markTestSkipped('SCRATCHBIRD_PHP_URL not set');
        }
        $pdo = new ScratchBirdPDO($dsn);
        $stmt = $pdo->query('SELECT * FROM sb_conformance.type_coverage');
        $row = $stmt->fetch(\PDO::FETCH_NUM);
        $this->assertNotFalse($row);
    }

    public function testCancel(): void
    {
        $dsn = getenv('SCRATCHBIRD_PHP_URL');
        if (!$dsn) {
            $this->markTestSkipped('SCRATCHBIRD_PHP_URL not set');
        }
        $cancelSql = getenv('SCRATCHBIRD_PHP_CANCEL_SQL');
        if (!$cancelSql) {
            $this->markTestSkipped('SCRATCHBIRD_PHP_CANCEL_SQL not set');
        }
        $conn = new \ScratchBird\PDO\Connection($dsn);
        $stream = $conn->executeQuery($cancelSql);
        $conn->cancel();
        $this->expectException(\Throwable::class);
        $stream->readRow();
    }
}

File: rust/tests/integration_test.rs
// ScratchBird-driver
// Copyright (c) 2025-2026 Dalton Calford
//
// Licensed under the Initial Developer's Public License Version 1.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at:
// https://www.firebirdsql.org/en/initial-developer-s-public-license-version-1-0/
use std::env;
use std::time::Duration;

use scratchbird::{Client, Config};
use scratchbird::types::{Param, Value};
use scratchbird::sql::Params;
use tokio::time::timeout;

#[tokio::test]
async fn basic_query() {
    let dsn = match env::var("SCRATCHBIRD_RUST_URL") {
        Ok(val) => val,
        Err(_) => return,
    };
    let config = Config::from_dsn(&dsn).unwrap();
    let mut client = Client::new(config);
    client.connect().await.unwrap();
    let result = client.query("SELECT 1").await.unwrap();
    client.close().await;
    assert!(!result.rows.is_empty());
}

#[tokio::test]
async fn prepare_bind_query() {
    let dsn = match env::var("SCRATCHBIRD_RUST_URL") {
        Ok(val) => val,
        Err(_) => return,
    };
    let config = Config::from_dsn(&dsn).unwrap();
    let mut client = Client::new(config);
    client.connect().await.unwrap();
    let result = client
        .query_params("SELECT ?::INTEGER", Params::from(vec![Param::Int32(42)]))
        .await
        .unwrap();
    client.close().await;
    assert!(!result.rows.is_empty());
    match result.rows[0][0] {
        Value::Int32(v) => assert_eq!(v, 42),
        Value::Int64(v) => assert_eq!(v, 42),
        _ => panic!("unexpected value type"),
    }
}

#[tokio::test]
async fn types_fixture_query() {
    let dsn = match env::var("SCRATCHBIRD_RUST_URL") {
        Ok(val) => val,
        Err(_) => return,
    };
    let config = Config::from_dsn(&dsn).unwrap();
    let mut client = Client::new(config);
    client.connect().await.unwrap();
    let result = client.query("SELECT * FROM sb_conformance.type_coverage").await.unwrap();
    client.close().await;
    assert!(!result.rows.is_empty());
}

#[tokio::test]
async fn cancel_query() {
    let dsn = match env::var("SCRATCHBIRD_RUST_URL") {
        Ok(val) => val,
        Err(_) => return,
    };
    let cancel_sql = match env::var("SCRATCHBIRD_RUST_CANCEL_SQL") {
        Ok(val) => val,
        Err(_) => return,
    };
    let config = Config::from_dsn(&dsn).unwrap();
    let mut client = Client::new(config);
    client.connect().await.unwrap();
    let result = timeout(Duration::from_millis(200), client.query(&cancel_sql)).await;
    if result.is_ok() {
        client.close().await;
        panic!("expected cancel timeout");
    }
    let _ = client.cancel().await;
    client.close().await;
}
